![JohnSnowLabs](https://nlp.johnsnowlabs.com/assets/images/logo.png)

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/JohnSnowLabs/spark-nlp-workshop/blob/master/healthcare-nlp/07.2.Task_Based_Clinical_Pretrained_Pipelines.ipynb)

# Task Based Clinical Pretrained Pipelines

## Colab Setup

In [ ]:
# Install the johnsnowlabs library to access Spark-OCR and Spark-NLP for Healthcare, Finance, and Legal.
! pip install -q johnsnowlabs

In [ ]:
from google.colab import files
print('Please Upload your John Snow Labs License using the button below')
license_keys = files.upload()

In [ ]:
from johnsnowlabs import nlp, medical
# After uploading your license run this to install all licensed Python Wheels and pre-download Jars the Spark Session JVM

nlp.install()

In [ ]:
from johnsnowlabs import nlp, medical
import pandas as pd
from pyspark.sql import functions as F

# Automatically load license data and start a session with all jars user has access to
spark = nlp.start()

👌 Detected license file /content/5.3.1.spark_nlp_for_healthcare.json
👷 Trying to install compatible secrets. Use nlp.settings.enforce_versions=False if you want to install outdated secrets.
👌 Launched cpu optimized session with with: 🚀Spark-NLP==5.3.1, 💊Spark-Healthcare==5.3.0, running on ⚡ PySpark==3.4.0


In [ ]:
spark

# Task Based Pretrained Pipelines


|index|pipeline|description|
|-----:|:-----|:-----|
| 1| [explain_clinical_doc_generic](https://nlp.johnsnowlabs.com/2024/01/17/explain_clinical_doc_generic_en.html)  |This pipeline is designed to extract all clinical/medical entities,<br> assign assertion status to the extracted entities,<br> establish relations between the extracted entities from the clinical texts.|
| 2| [explain_clinical_doc_granular](https://nlp.johnsnowlabs.com/2024/01/26/explain_clinical_doc_granular_en.html)  |This pipeline is designed to extract all clinical/medical entities,<br> assign assertion status to the extracted entities,<br> establish relations between the extracted entities from the clinical texts.|
|3| [explain_clinical_doc_biomarker](https://nlp.johnsnowlabs.com/2024/03/11/explain_clinical_doc_biomarker_en.html)  | This specialized biomarker pipeline can extract biomarker entities,<br>classify sentences whether they contain biomarker entities or not,<br>establish relations between the extracted biomarker and biomarker<br>results from the clinical documents.|
4| [explain_clinical_doc_oncology](https://nlp.johnsnowlabs.com/2024/01/18/explain_clinical_doc_oncology_en.html)  |This specialized oncology pipeline can extract oncological entities,<br>  assign assertion status to the extracted entities, <br> establish relations between the extracted entities from the clinical documents.|
| 5| [explain_clinical_doc_radiology](https://nlp.johnsnowlabs.com/2024/01/18/explain_clinical_doc_radiology_en.html) |This pipeline is designed to extract all clinical/medical entities,<br> assign assertion status to the extracted entities,<br> establish relations between the extracted entities from the clinical texts.|
| 6| [explain_clinical_doc_vop](https://nlp.johnsnowlabs.com/2024/01/16/explain_clinical_doc_vop_en.html)|This pipeline is designed to extract healthcare-related terms entities,<br> assign assertion status to the extracted entities, establish relations<br> between the extracted entities from the documents transferred from the patient’s sentences.|
| 7| [explain_clinical_doc_carp](https://nlp.johnsnowlabs.com/2023/06/17/explain_clinical_doc_carp_en.html) |A pipeline with `ner_clinical`, `assertion_dl`, `re_clinical` and `ner_posology`. <br>It will extract clinical and medication entities, <br>assign assertion status and find relationships between clinical entities.|
| 8| [explain_clinical_doc_era](https://nlp.johnsnowlabs.com/2023/06/17/explain_clinical_doc_era_en.html)  |A pipeline with `ner_clinical_events`, `assertion_dl` and `re_temporal_events_clinical`. <br>It will extract clinical entities, assign assertion status and find <br>temporal relationships between clinical entities.|
| 9| [explain_clinical_doc_ade](https://nlp.johnsnowlabs.com/2023/06/17/explain_clinical_doc_ade_en.html)|A pipeline for Adverse Drug Events (ADE) with `ner_ade_biobert`, `assertion_dl_biobert`, `classifierdl_ade_conversational_biobert`, <br>and `re_ade_biobert` . It will classify the document, extract ADE and DRUG clinical<br> entities, assign assertion status to ADE entities, and relate Drugs with their ADEs.|
| 10| [explain_clinical_doc_medication](https://nlp.johnsnowlabs.com/2023/06/17/explain_clinical_doc_medication_en.html)|A pipeline for detecting posology entities with the `ner_posology_large` <br>NER model, assigning their assertion status with `assertion_jsl` model<br>, and extracting relations between posology-related terminology<br> with `posology_re` relation extraction model.|
| 11| [explain_clinical_doc_risk_factors](https://nlp.johnsnowlabs.com/2024/03/25/explain_clinical_doc_risk_factors_en.html)|This pipeline is designed to extract all clinical/medical entities,<br> which may be considered as risk factors from text, <br>assign assertion status to the extracted entities, establish relations between the extracted entities.|
| 12| [explain_clinical_doc_public_health](https://nlp.johnsnowlabs.com/2024/03/19/explain_clinical_doc_public_health_en.html)|This specialized public health pipeline can extract public health-related entities,<br> assign assertion status to the extracted entities,<br> establish relations between the extracted entities from the clinical documents. In this pipeline, five NER, one assertion, <br>and one relation extraction model were used to achieve those tasks.|

### explain_clinical_doc_granular

A pipeline with ner_jsl, assertion_jsl and re_test_result_date. It will extract clinical and medication entities, assign assertion status and find relationships between clinical entities.

- Clinical Entity Labels: `Admission_Discharge`, `Age`, `Alcohol`, `Allergen`, `BMI`, `Birth_Entity`, `Blood_Pressure`, `Cerebrovascular_Disease`, `Clinical_Dept`, `Communicable_Disease`, `Date`, `Death_Entity`, `Diabetes`, `Diet`, `Direction`, `Disease_Syndrome_Disorder`, `Dosage`, `Drug_BrandName`, `Drug_Ingredient`, `Duration`, `EKG_Findings`, `Employment`, `External_body_part_or_region`, `Family_History_Header`, `Fetus_NewBorn`, `Form`, `Frequency`, `Gender`, `HDL`, `Heart_Disease`, `Height`, `Hyperlipidemia`, `Hypertension`, `ImagingFindings`, `Imaging_Technique`, `Injury_or_Poisoning`, `Internal_organ_or_component`, `Kidney_Disease`, `LDL`, `Labour_Delivery`, `Medical_Device`, `Medical_History_Header`, `Modifier`, `O2_Saturation`, `Obesity`, `Oncological`, `Overweight`, `Oxygen_Therapy`, `Pregnancy`, `Procedure`, `Psychological_Condition`, `Pulse`, `Race_Ethnicity`, `Relationship_Status`, `RelativeDate`, `RelativeTime`, `Respiration`, `Route`, `Section_Header`, `Sexually_Active_or_Sexual_Orientation`, `Smoking`, `Social_History_Header`, `Strength`, `Substance`, `Substance_Quantity`, `Symptom`, `Temperature`, `Test`, `Test_Result`, `Time`, `Total_Cholesterol`, `Treatment`, `Triglycerides`, `VS_Finding`, `Vaccine`, `Vaccine_Name`, `Vital_Signs_Header`, `Weight`

- Assertion Status Labels: `Hypothetical`, `Someoneelse`, `Past`, `Absent`, `Family`, `Planned`, `Possible`, `Present`

- Relation Extraction Labels: `is_finding_of`, `is_date_of`, `is_result_of`, `O`


In [ ]:
# returns relation pair dataframe
import pandas as pd

def get_relations_df (results, col='relations'):
  rel_pairs=[]
  for rel in results[0][col]:
      rel_pairs.append((
          rel.result,
          rel.metadata['entity1'], rel.metadata['entity1_begin'], rel.metadata['entity1_end'], rel.metadata['chunk1'],
          rel.metadata['entity2'], rel.metadata['entity2_begin'], rel.metadata['entity2_end'], rel.metadata['chunk2'],
          rel.metadata['confidence']
      ))

  rel_df = pd.DataFrame(rel_pairs, columns=['relation',
                                            'entity1','entity1_begin','entity1_end','chunk1',
                                            'entity2','entity2_begin','entity2_end','chunk2',
                                            'confidence'])

  rel_df.confidence = rel_df.confidence.astype(float)

  return rel_df

In [ ]:
from sparknlp.pretrained import PretrainedPipeline

In [ ]:
pipeline = nlp.PretrainedPipeline("explain_clinical_doc_granular", "en", "clinical/models")

explain_clinical_doc_granular download started this may take some time.
Approx size to download 1.6 GB
[OK!]


In [ ]:
pipeline.model.stages

[DocumentAssembler_27dc46af8fb2,
 SentenceDetectorDLModel_6bafc4746ea5,
 REGEX_TOKENIZER_b17b2f22704f,
 WORD_EMBEDDINGS_MODEL_9004b1d00302,
 MedicalNerModel_c89cbceb1028,
 NER_CONVERTER_59ce0f445e62,
 NER_CONVERTER_256c8eb7a1be,
 ASSERTION_DL_e5e007602386,
 POS_6f55785005bf,
 dependency_e7755462ba78,
 RelationExtractionModel_ce79d77d1bf1]

#### with fullAnnotate()

In [ ]:
text ="""The patient admitted for gastrointestinal pathology, under working treatment.
History of prior heart murmur with echocardiogram findings as above on March 1998.
According to the latest echocardiogram, basically revealed normal left ventricular function with left atrial enlargement .
Based on the above findings, we will treat her medically with ACE inhibitors and diuretics and see how she fares.
"""

result = pipeline.fullAnnotate(text)[0]

In [ ]:
result.keys()

dict_keys(['assertion_ner_chunk', 'document', 'jsl_ner_chunk', 'assertion', 'jsl_ner', 'token', 'relations', 'embeddings', 'pos_tags', 'dependencies', 'sentence'])

**NER Results**

In [ ]:
import pandas as pd
chunks=[]
entities=[]
begins=[]
ends=[]

for n in result['jsl_ner_chunk']:

    chunks.append(n.result)
    begins.append(n.begin)
    ends.append(n.end)
    entities.append(n.metadata['entity'])

df = pd.DataFrame({'chunks':chunks, 'begin':begins, 'end':ends, 'entities':entities})

df

chunks  begin  end             entities
0                     admitted     12   19  Admission_Discharge
1   gastrointestinal pathology     25   50        Clinical_Dept
2                 heart murmur     95  106        Heart_Disease
3               echocardiogram    113  126                 Test
4                   March 1998    149  158                 Date
5               echocardiogram    185  198                 Test
6                       normal    220  225          Test_Result
7    left ventricular function    227  251                 Test
8      left atrial enlargement    258  280        Heart_Disease
9                          her    327  329               Gender
10              ACE inhibitors    346  359      Drug_Ingredient
11                   diuretics    365  373      Drug_Ingredient
12                         she    387  389               Gender

**Assertion Status Results**

In [ ]:
chunks=[]
entities=[]
status=[]

for n,m in zip(result['assertion_ner_chunk'],result['assertion']):

    chunks.append(n.result)
    entities.append(n.metadata['entity'])
    status.append(m.result)

df = pd.DataFrame({'chunks':chunks, 'entities':entities, 'assertion':status})

df

chunks         entities assertion
0               heart murmur    Heart_Disease      Past
1             echocardiogram             Test      Past
2             echocardiogram             Test   Present
3                     normal      Test_Result   Present
4  left ventricular function             Test   Present
5    left atrial enlargement    Heart_Disease   Present
6             ACE inhibitors  Drug_Ingredient   Planned
7                  diuretics  Drug_Ingredient   Planned

**Relation Extraction Results**

In [ ]:
annotations = pipeline.fullAnnotate(text)

rel_df = get_relations_df (annotations, 'relations')

rel_df[rel_df.relation!="O"]

relation        entity1 entity1_begin entity1_end  \
0  is_finding_of  Heart_Disease            95         106   
1     is_date_of  Heart_Disease            95         106   
2     is_date_of           Test           113         126   
4  is_finding_of           Test           185         198   
5   is_result_of    Test_Result           220         225   
6  is_finding_of           Test           227         251   

                      chunk1        entity2 entity2_begin entity2_end  \
0               heart murmur           Test           113         126   
1               heart murmur           Date           149         158   
2             echocardiogram           Date           149         158   
4             echocardiogram  Heart_Disease           258         280   
5                     normal           Test           227         251   
6  left ventricular function  Heart_Disease           258         280   

                      chunk2  confidence  
0             echocardiogram         1.0  
1                 March 1998         1.0  
2                 March 1998         1.0  
4    left atrial enlargement         1.0  
5  left ventricular function         1.0  
6    left atrial enlargement         1.0

#### with transform()

In [ ]:
data = spark.createDataFrame([["""
The patient admitted for gastrointestinal pathology, under working treatment.
History of prior heart murmur with echocardiogram findings as above on March 1998.
According to the latest echocardiogram, basically revealed normal left ventricular function with left atrial enlargement .
Based on the above findings, we will treat her medically with ACE inhibitors and diuretics and see how she fares.
"""]]).toDF("text")

result = pipeline.transform(data)

result.select(F.explode(F.arrays_zip(result.jsl_ner_chunk.result,
                                     result.jsl_ner_chunk.begin,
                                     result.jsl_ner_chunk.end,
                                     result.jsl_ner_chunk.metadata)).alias("cols"))\
        .select(F.expr("cols['0']").alias("chunk"),
                F.expr("cols['1']").alias("begin"),
                F.expr("cols['2']").alias("end"),
                F.expr("cols['3']['entity']").alias("ner_label")).show()

+--------------------+-----+---+-------------------+
|               chunk|begin|end|          ner_label|
+--------------------+-----+---+-------------------+
|            admitted|   13| 20|Admission_Discharge|
|gastrointestinal ...|   26| 51|      Clinical_Dept|
|        heart murmur|   96|107|      Heart_Disease|
|      echocardiogram|  114|127|               Test|
|          March 1998|  150|159|               Date|
|      echocardiogram|  186|199|               Test|
|              normal|  221|226|        Test_Result|
|left ventricular ...|  228|252|               Test|
|left atrial enlar...|  259|281|      Heart_Disease|
|                 her|  328|330|             Gender|
|      ACE inhibitors|  347|360|    Drug_Ingredient|
|           diuretics|  366|374|    Drug_Ingredient|
|                 she|  388|390|             Gender|
+--------------------+-----+---+-------------------+



In [ ]:
result.select(F.explode(F.arrays_zip(result.assertion_ner_chunk.result,
                                     result.assertion_ner_chunk.begin,
                                     result.assertion_ner_chunk.end,
                                     result.assertion_ner_chunk.metadata,
                                     result.assertion.result)).alias("cols"))\
        .select(F.expr("cols['0']").alias("chunk"),
                F.expr("cols['1']").alias("begin"),
                F.expr("cols['2']").alias("end"),
                F.expr("cols['3']['entity']").alias("ner_label"),
                F.expr("cols['4']").alias("assertion")).show()

+--------------------+-----+---+---------------+---------+
|               chunk|begin|end|      ner_label|assertion|
+--------------------+-----+---+---------------+---------+
|        heart murmur|   96|107|  Heart_Disease|     Past|
|      echocardiogram|  114|127|           Test|     Past|
|      echocardiogram|  186|199|           Test|  Present|
|              normal|  221|226|    Test_Result|  Present|
|left ventricular ...|  228|252|           Test|  Present|
|left atrial enlar...|  259|281|  Heart_Disease|  Present|
|      ACE inhibitors|  347|360|Drug_Ingredient|  Planned|
|           diuretics|  366|374|Drug_Ingredient|  Planned|
+--------------------+-----+---+---------------+---------+



In [ ]:
result.select(F.explode(F.arrays_zip(result.token.result,
                                     result.jsl_ner.result,
                                     result.pos_tags.result,
                                     result.dependencies.result)).alias("cols"))\
        .select(F.expr("cols['0']").alias("tokens"),
                F.expr("cols['1']").alias("jsl_ner"),
                F.expr("cols['2']").alias("pos_tags"),
                F.expr("cols['3']").alias("dependencies")).show()

+----------------+--------------------+--------+------------+
|          tokens|             jsl_ner|pos_tags|dependencies|
+----------------+--------------------+--------+------------+
|             The|                   O|      DD|    admitted|
|         patient|                   O|      NN|    admitted|
|        admitted|B-Admission_Disch...|     VVN|        ROOT|
|             for|                   O|      II|   pathology|
|gastrointestinal|     B-Clinical_Dept|      JJ|   pathology|
|       pathology|     I-Clinical_Dept|      NN|    admitted|
|               ,|                   O|      NN|    admitted|
|           under|                   O|      II|   treatment|
|         working|                   O|    VVGJ|   treatment|
|       treatment|                   O|      NN|    admitted|
|               .|                   O|      NN|    admitted|
|         History|                   O|      NN|        ROOT|
|              of|                   O|      II|       heart|
|       

# Clinical Deidentification

This pipeline can be used to deidentify PHI information from medical texts. The PHI information will be masked and obfuscated in the resulting text. The pipeline can mask and obfuscate  entities.

|index|model|lang|index|model|lang|
|-----:|:-----|----|-----:|:-----|----|
| 1| [clinical_deidentification](https://nlp.johnsnowlabs.com/2022/09/14/clinical_deidentification_en.html)  |ar, de, en, es, fr, it, pt, ro| 9| [clinical_deidentification_obfuscation_medium](https://nlp.johnsnowlabs.com/2024/02/09/clinical_deidentification_obfuscation_medium_en.html)  |en|
| 2| [clinical_deidentification_augmented](https://nlp.johnsnowlabs.com/2022/03/03/clinical_deidentification_augmented_es_2_4.html)  |es| 10| [clinical_deidentification_obfuscation_small](https://nlp.johnsnowlabs.com/2024/02/09/clinical_deidentification_obfuscation_small_en.html)  |en|
| 3| [clinical_deidentification_generic](https://nlp.johnsnowlabs.com/2024/02/21/clinical_deidentification_generic_en.html)  |en| 11| [clinical_deidentification_slim](https://nlp.johnsnowlabs.com/2023/06/17/clinical_deidentification_slim_en.html)  |en|
| 4| [clinical_deidentification_generic_optimized](https://nlp.johnsnowlabs.com/2024/03/14/clinical_deidentification_generic_optimized_en.html)  |en| 12| [clinical_deidentification_subentity](https://nlp.johnsnowlabs.com/2024/02/21/clinical_deidentification_subentity_en.html)  |en|
| 5| [clinical_deidentification_glove](https://nlp.johnsnowlabs.com/2022/03/04/clinical_deidentification_glove_en_3_0.html)  |en| 13| [clinical_deidentification_subentity_nameAugmented](https://nlp.johnsnowlabs.com/2024/03/14/clinical_deidentification_subentity_nameAugmented_en.html)  |en|
| 6| [clinical_deidentification_glove_augmented](https://nlp.johnsnowlabs.com/2022/09/16/clinical_deidentification_glove_augmented_en.html)  |en| 14| [clinical_deidentification_subentity_optimized](https://nlp.johnsnowlabs.com/2024/03/14/clinical_deidentification_subentity_optimized_en.html)  |en|
| 7| [clinical_deidentification_langtest](https://nlp.johnsnowlabs.com/2024/01/10/clinical_deidentification_langtest_en.html)  |en| 15| [clinical_deidentification_wip](https://nlp.johnsnowlabs.com/2023/06/17/clinical_deidentification_wip_en.html)  |en|
| 8| [clinical_deidentification_multi_mode_output](https://nlp.johnsnowlabs.com/2024/03/27/clinical_deidentification_multi_mode_output_en.html)  |en|

You can find **`German`, `Spanish`, `French`, `Italian`, `Portuguese`, `Romanian`**  and **`Arabic`**  deidentification models and pretrained pipeline examples in this notebook:   [Clinical Multi Language Deidentification Notebook](https://github.com/JohnSnowLabs/spark-nlp-workshop/blob/master/tutorials/Certification_Trainings/Healthcare/4.1.Clinical_Multi_Language_Deidentification.ipynb)

We have also NER pipelines:

|index|model|lang|index|model|lang|
|-----:|:-----|----|-----:|:-----|----|
| 1| [ner_deid_augmented_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_augmented_pipeline_en_3_0.html)  |en| 10| [ner_deid_sd_large_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_sd_large_pipeline_en_3_0.html)  |en|
| 2| [ner_deid_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_biobert_pipeline_en_3_0.html)  |en| 11| [ner_deid_sd_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_sd_pipeline_en_3_0.html)  |en|
| 3| [ner_deid_enriched_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_enriched_biobert_pipeline_en_3_0.html)  |en| 12| [ner_deid_subentity_augmented_i2b2_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_subentity_augmented_i2b2_pipeline_en_3_0.html)  |en|
| 4| [ner_deid_enriched_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_enriched_pipeline_en_3_0.html)  |en| 13| [ner_deid_subentity_augmented_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_subentity_augmented_pipeline_en_3_0.html)  |en|
| 5| [ner_deid_generic_augmented_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_generic_augmented_pipeline_en_3_0.html)  |en| 14| [ner_deid_subentity_bert_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_deid_subentity_bert_pipeline_ro.html)  |ro|
| 6| [ner_deid_generic_bert_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_deid_generic_bert_pipeline_ro.html)  |ro| 15| [ner_deid_subentity_glove_pipeline](https://nlp.johnsnowlabs.com/2023/03/13/ner_deid_subentity_glove_pipeline_en.html)  |en|
| 7| [ner_deid_generic_glove_pipeline](https://nlp.johnsnowlabs.com/2023/03/13/ner_deid_generic_glove_pipeline_en.html)  |en| 16| [ner_deid_subentity_pipeline](https://nlp.johnsnowlabs.com/2023/05/31/ner_deid_subentity_pipeline_ar.html)  |ar, de, it, ro|
| 8| [ner_deid_generic_pipeline](https://nlp.johnsnowlabs.com/2023/05/31/ner_deid_generic_pipeline_ar.html)  |ar, de, it, ro| 17| [ner_deid_synthetic_pipeline](https://nlp.johnsnowlabs.com/2023/03/13/ner_deid_synthetic_pipeline_en.html)  |en|
| 9| [ner_deid_large_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deid_large_pipeline_en_3_0.html)  |en| 18| [ner_deidentify_dl_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_deidentify_dl_pipeline_en_3_0.html)  |en|

### clinical_deidentification

In [ ]:
deid_pipeline = nlp.PretrainedPipeline("clinical_deidentification", "en", "clinical/models")

clinical_deidentification download started this may take some time.
Approx size to download 1.6 GB
[OK!]


In [ ]:
deid_res = deid_pipeline.annotate("Record date : 2093-01-13 , David Hale , M.D .  Name : Hendrickson , Ora MR 25 years-old . # 719435 Date : 01/13/93 . Signed by Oliveira Sander . Record date : 2079-11-09 . Cocke County Baptist Hospital . 0295 Keats Street. Phone 302-786-5227.")

In [ ]:
deid_res.keys()

dict_keys(['masked', 'obfuscated', 'ner_chunk', 'masked_fixed_length_chars', 'sentence', 'masked_with_chars'])

In [ ]:
pd.set_option("display.max_colwidth", 100)

df= pd.DataFrame(list(zip(deid_res["sentence"],
                          deid_res["masked"],
                          deid_res["masked_with_chars"],
                          deid_res["masked_fixed_length_chars"],
                          deid_res["obfuscated"])),
                 columns= ["Sentence", "Masked", "Masked with Chars", "Masked with Fixed Chars", "Obfuscated"])

df

Sentence  \
0       Record date : 2093-01-13 , David Hale , M.D .   
1          Name : Hendrickson , Ora MR 25 years-old .   
2                          # 719435 Date : 01/13/93 .   
3                         Signed by Oliveira Sander .   
4                          Record date : 2079-11-09 .   
5  Cocke County Baptist Hospital . 0295 Keats Street.   
6                                 Phone 302-786-5227.   

                                    Masked  \
0  Record date : <DATE> , <DOCTOR> , M.D .   
1    Name : <PATIENT> MR <AGE> years-old .   
2                # <PHONE> Date : <DATE> .   
3                     Signed by <DOCTOR> .   
4                   Record date : <DATE> .   
5                              <LOCATION>.   
6                           Phone <PHONE>.   

                                    Masked with Chars  \
0       Record date : [********] , [********] , M.D .   
1          Name : [***************] MR ** years-old .   
2                          # [****] Date : [******] .   
3                         Signed by [*************] .   
4                          Record date : [********] .   
5  [***********************************************].   
6                                 Phone [**********].   

             Masked with Fixed Chars  \
0  Record date : **** , **** , M.D .   
1    Name : **** MR **** years-old .   
2               # **** Date : **** .   
3                   Signed by **** .   
4               Record date : **** .   
5                              ****.   
6                        Phone ****.   

                                        Obfuscated  
0  Record date : 2093-02-02 , Layne Benton , M.D .  
1            Name : Toma Deiters MR 39 years-old .  
2                       # 161096 Date : 02/02/93 .  
3                           Signed by Roxy Manns .  
4                       Record date : 2079-11-29 .  
5                             11700 N Meridian St.  
6                              Phone 045-409-8119.

# Resolver Pipelines

We have **Resolver pipelines** for converting clinical entities to their UMLS CUI codes. You will just feed your text and it will return the corresponding UMLS codes.



**Resolver Pipelines:**

|index|pipeline|Entity|Target
|-:|:-|:-|:-|
| 1|[umls_disease_syndrome_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/07/26/umls_disease_syndrome_resolver_pipeline_en_3_0.html)|Disease and Syndromes|UMLS CUI|
| 2|[umls_drug_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/07/26/umls_drug_resolver_pipeline_en_3_0.html)|Drug|UMLS CUI|
| 3|[umls_drug_substance_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/07/25/umls_drug_substance_resolver_pipeline_en_3_0.html)|Drug Substance|UMLS CUI|
| 4|[medication_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/09/01/medication_resolver_pipeline_en.html)|Drug|Adverse Reaction, RxNorm, UMLS<br>NDC, SNOMED CT|
| 5|[medication_resolver_transform_pipeline](https://nlp.johnsnowlabs.com/2022/09/01/medication_resolver_transform_pipeline_en.html)|Drug|Adverse Reaction, RxNorm<br>UMLS, NDC, SNOMED CT|
| 6|[umls_major_concepts_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/07/25/umls_major_concepts_resolver_pipeline_en_3_0.html)|Clinical Major Concepts| UMLS CUI|
| 7|[umls_clinical_findings_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/07/26/umls_clinical_findings_resolver_pipeline_en_3_0.html)|Clinical Findings|	UMLS CUI|
| 8|[icd9_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/09/30/icd9_resolver_pipeline_en.html)|Clinical Findings|ICD-9|
| 9|[atc_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/17/atc_resolver_pipeline_en.html)|Drug|ATC|
| 10|[cpt_procedures_measurements_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/17/cpt_procedures_measurements_resolver_pipeline_en.html)|Procedure, Measurement|CPT|
| 11|[hcc_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/17/hcc_resolver_pipeline_en.html)|Clinical Findings|HCC|
| 12|[hpo_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/17/hpo_resolver_pipeline_en.html)|Human Phenotype|HPO|
| 13|[cvx_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/10/12/cvx_resolver_pipeline_en.html)|Vaccine|CVX|
| 14|[icd10cm_resolver_pipeline](https://nlp.johnsnowlabs.com/2022/11/02/icd10cm_resolver_pipeline_en.html)|Problem|ICD-10|
| 15|[icd10pcs_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/02/02/icd10pcs_resolver_pipeline_en.html)|Procedure|ICD-10-PCS|
| 16|[icdo_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/02/02/icdo_resolver_pipeline_en.html)|Problem|ICD-O|
| 17|[loinc_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/02/02/loinc_resolver_pipeline_en.html)|Test|LOINC|
| 17|[loinc_numeric_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/30/loinc_numeric_resolver_pipeline_en.html)|Test|LOINC|
| 18|[snomed_body_structure_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/17/snomed_body_structure_resolver_pipeline_en.html)|Clinical Findings|SNOMED|
| 19|[snomed_findings_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/17/snomed_findings_resolver_pipeline_en.html)|Clinical Findings|SNOMED CT|
| 20|[snomed_procedures_measurements_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/31/snomed_findings_resolver_pipelin_en.html)|Procedure, Test|SNOMED|
| 21|[mesh_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/25/mesh_resolver_pipeline_en.html)|Clinical Findings|MeSH|
| 22|[ndc_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/25/ndc_resolver_pipeline_en.html)|Drug|NDC|
| 23|[ncit_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/25/ncit_resolver_pipeline_en.html)|Clinical Findings|NCIt|
| 24|[rxcui_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/02/01/rxcui_resolver_pipeline_en.html)|Clinical Findings|RxCUI|
| 25|[hcpcs_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/30/hcpcs_resolver_pipeline_en.html)|Procedure|HCPCS|
| 26|[hgnc_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/30/hgnc_resolver_pipeline_en.html)|Gene|HGNC|
| 27|[icd10cm_generalised_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/01/30/icd10cm_generalised_resolver_pipeline_en.html)|Clinical Findings|ICD-10-CM|
| 28|[abbreviation_pipeline](https://nlp.johnsnowlabs.com/2023/08/16/abbreviation_pipeline_en.html)|ABBR|Definitions and Categories|
| 29|[icd10cm_multi_mapper_pipeline](https://nlp.johnsnowlabs.com/2023/08/16/icd10cm_multi_mapper_pipeline_en.html)|ICD-10-CM|Billable Mappings, HCC Codes<br>Cause Mappings, Claim Mappings<br>SNOMED Codes, UMLS Codes<br> ICD-9 Codes|
| 30|[rxnorm_multi_mapper_pipeline](https://nlp.johnsnowlabs.com/2023/08/16/rxnorm_multi_mapper_pipeline_en.html)|RxNorm|Drug Brand Names, RxNorm Extension Brand Names<br> Action Mappings, Treatment Mappings<br> UMLS Codes, NDC Product Codes<br> NDC Package Codes|
| 31|[rxnorm_resolver_pipeline](https://nlp.johnsnowlabs.com/2023/08/16/rxnorm_resolver_pipeline_en.html)|Drug|RxNorm |
| 32|[snomed_multi_mapper_pipeline](https://nlp.johnsnowlabs.com/2023/08/16/snomed_multi_mapper_pipeline_en.html)|SNOMED Codes|ICD-10, ICD-O,UMLS |
| 33|[icd10cm_rxnorm_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/03/07/icd10cm_rxnorm_resolver_pipeline_en.html)|Drug|ICD-10, RxNorm |
| 34|[snomed_term_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/03/22/snomed_term_resolver_pipeline_en.html)|SNOMED Codes|SNOMED terms |
| 35|[snomed_auxConcepts_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/03/12/snomed_auxConcepts_resolver_pipeline_en.html)|Morph Abnormality, Clinical Drug<br>Clinical Drug Form, Procedure<br> Substance, Physical Object<br> Body Structure|SNOMED Codes |
| 36|[snomed_conditions_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/03/11/snomed_conditions_resolver_pipeline_en.html)|Clinical Findings|SNOMED Codes |
| 37|[snomed_drug_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/03/11/snomed_drug_resolver_pipeline_en.html)|Drug|SNOMED Codes |
| 38|[snomed_resolver_pipeline](https://nlp.johnsnowlabs.com/2024/03/11/snomed_resolver_pipeline_en.html)|Clinical Findings, Morph Abnormality<br>Clinical Drug, Clinical Drug Form<br> Procedure, Substance<br>Physical Object, Body Structure|SNOMED Codes |

###**icd10cm_resolver_pipeline**

This pretrained pipeline extracts clinical conditions and map them to their corresponding ICD-10-CM codes. You’ll just feed your text and it will detect the related entities and return their corresponding ICD-10-CM codes.

In [ ]:
resolver_pipeline = nlp.PretrainedPipeline("icd10cm_resolver_pipeline", "en", "clinical/models")

text = """A 28-year-old female with a history of gestational diabetes mellitus diagnosed eight years and anisakiasis. Also, it was reported that fetal and neonatal hemorrhage"""

result = resolver_pipeline.fullAnnotate(text)[0]

icd10cm_resolver_pipeline download started this may take some time.
Approx size to download 3.3 GB
[OK!]


In [ ]:
import pandas as pd
chunks=[]
entities=[]
icd10cm_code=[]


for n,m in zip(result['chunk'], result['icd10cm']):

    chunks.append(n.result)
    entities.append(n.metadata['entity'])
    icd10cm_code.append(m.result)


df = pd.DataFrame({'chunks':chunks,
                   'entities':entities,
                   'icd10cm_code':icd10cm_code})

df

chunks entities icd10cm_code
0  gestational diabetes mellitus  PROBLEM      O24.919
1                    anisakiasis  PROBLEM        B81.0
2  fetal and neonatal hemorrhage  PROBLEM         P549

# NER Pipelines

**`NER pretrained ` Model List**

|index|model|index|model|index|model|index|model|
|-----:|:-----|-----:|:-----|-----:|:-----|-----:|:-----|
| 1| [jsl_ner_wip_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/jsl_ner_wip_clinical_pipeline_en_3_0.html)  | 2| [jsl_ner_wip_greedy_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/jsl_ner_wip_greedy_biobert_pipeline_en_3_0.html)  | 3| [jsl_ner_wip_greedy_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/jsl_ner_wip_greedy_clinical_pipeline_en_3_0.html)  | 4| [jsl_ner_wip_modifier_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/jsl_ner_wip_modifier_clinical_pipeline_en_3_0.html)  |
| 5| [jsl_rd_ner_wip_greedy_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/jsl_rd_ner_wip_greedy_biobert_pipeline_en_3_0.html)  | 6| [jsl_rd_ner_wip_greedy_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/jsl_rd_ner_wip_greedy_clinical_pipeline_en_3_0.html)  | 7| [ner_abbreviation_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_abbreviation_clinical_pipeline_en_3_0.html)  | 8| [ner_ade_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_ade_biobert_pipeline_en_3_0.html)  |
| 9| [ner_ade_clinical_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_ade_clinical_langtest_pipeline_en.html)  | 10| [ner_ade_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_ade_clinical_pipeline_en_3_0.html)  | 11| [ner_ade_clinicalbert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_ade_clinicalbert_pipeline_en_3_0.html)  | 12| [ner_ade_healthcare_pipeline](https://nlp.johnsnowlabs.com/2022/03/22/ner_ade_healthcare_pipeline_en_3_0.html)  |
| 13| [ner_anatomy_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_anatomy_biobert_pipeline_en_3_0.html)  | 14| [ner_anatomy_coarse_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_anatomy_coarse_biobert_pipeline_en_3_0.html)  | 15| [ner_anatomy_coarse_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_anatomy_coarse_pipeline_en_3_0.html)  | 16| [ner_anatomy_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_anatomy_pipeline_en_3_0.html)  |
| 17| [ner_bacterial_species_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_bacterial_species_pipeline_en_3_0.html)  | 18| [ner_biomarker_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_biomarker_pipeline_en_3_0.html)  | 19| [ner_biomedical_bc2gm_pipeline](https://nlp.johnsnowlabs.com/2023/03/14/ner_biomedical_bc2gm_pipeline_en.html)  | 20| [ner_bionlp_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_bionlp_biobert_pipeline_en_3_0.html)  |
| 21| [ner_bionlp_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_bionlp_pipeline_en_3_0.html)  | 22| [ner_cancer_genetics_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_cancer_genetics_pipeline_en_3_0.html)  | 23| [ner_cellular_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_cellular_biobert_pipeline_en_3_0.html)  | 24| [ner_cellular_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_cellular_pipeline_en_3_0.html)  |
| 25| [ner_chemd_clinical_pipeline](https://nlp.johnsnowlabs.com/2023/03/14/ner_chemd_clinical_pipeline_en.html)  | 26| [ner_chemicals_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_chemicals_pipeline_en_3_0.html)  | 27| [ner_chemprot_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_chemprot_biobert_pipeline_en_3_0.html)  | 28| [ner_chemprot_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_chemprot_clinical_pipeline_en_3_0.html)  |
| 29| [ner_chexpert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_chexpert_pipeline_en_3_0.html)  | 30| [ner_clinical_bert_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_clinical_bert_pipeline_ro.html)  | 31| [ner_clinical_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_clinical_biobert_pipeline_en_3_0.html)  | 32| [ner_clinical_large_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_clinical_large_pipeline_en_3_0.html)  |
| 33| [ner_clinical_pipeline](https://nlp.johnsnowlabs.com/2023/09/02/ner_clinical_pipeline_es.html)  | 34| [ner_clinical_trials_abstracts_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_clinical_trials_abstracts_pipeline_en.html)  | 35| [ner_covid_trials_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_covid_trials_pipeline_en.html)  | 36| [ner_diag_proc_pipeline](https://nlp.johnsnowlabs.com/2023/03/15/ner_diag_proc_pipeline_es.html)  |
| 37| [ner_diseases_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_diseases_biobert_pipeline_en_3_0.html)  | 38| [ner_diseases_large_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_diseases_large_pipeline_en_3_0.html)  | 39| [ner_diseases_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_diseases_pipeline_en_3_0.html)  | 40| [ner_drugprot_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_drugprot_clinical_pipeline_en_3_0.html)  |
| 41| [ner_drugs_greedy_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_drugs_greedy_pipeline_en_3_0.html)  | 42| [ner_drugs_large_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_drugs_large_pipeline_en_3_0.html)  | 43| [ner_drugs_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_drugs_pipeline_en_3_0.html)  | 44| [ner_eu_clinical_case_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_eu_clinical_case_pipeline_en.html)  |
| 45| [ner_eu_clinical_condition_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_eu_clinical_condition_pipeline_es.html)  | 46| [ner_events_admission_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_events_admission_clinical_pipeline_en_3_0.html)  | 47| [ner_events_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_events_biobert_pipeline_en_3_0.html)  | 48| [ner_events_clinical_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_events_clinical_langtest_pipeline_en.html)  |
| 49| [ner_events_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_events_clinical_pipeline_en_3_0.html)  | 50| [ner_events_healthcare_pipeline](https://nlp.johnsnowlabs.com/2022/03/22/ner_events_healthcare_pipeline_en_3_0.html)  | 51| [ner_genetic_variants_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_genetic_variants_pipeline_en_3_0.html)  | 52| [ner_healthcare_pipeline](https://nlp.johnsnowlabs.com/2022/03/22/ner_healthcare_pipeline_en_3_0.html)  |
| 53| [ner_healthcare_slim_pipeline](https://nlp.johnsnowlabs.com/2023/03/15/ner_healthcare_slim_pipeline_de.html)  | 54| [ner_human_phenotype_gene_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_human_phenotype_gene_biobert_pipeline_en_3_0.html)  | 55| [ner_human_phenotype_gene_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_human_phenotype_gene_clinical_pipeline_en_3_0.html)  | 56| [ner_human_phenotype_go_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_human_phenotype_go_biobert_pipeline_en_3_0.html)  |
| 57| [ner_human_phenotype_go_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_human_phenotype_go_clinical_pipeline_en_3_0.html)  | 58| [ner_jsl_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_jsl_biobert_pipeline_en_3_0.html)  | 59| [ner_jsl_enriched_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_jsl_enriched_biobert_pipeline_en_3_0.html)  | 60| [ner_jsl_enriched_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_jsl_enriched_pipeline_en_3_0.html)  |
| 61| [ner_jsl_greedy_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_jsl_greedy_biobert_pipeline_en_3_0.html)  | 62| [ner_jsl_greedy_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_jsl_greedy_pipeline_en_3_0.html)  | 63| [ner_jsl_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_jsl_langtest_pipeline_en.html)  | 64| [ner_jsl_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_jsl_pipeline_en_3_0.html)  |
| 65| [ner_jsl_slim_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_jsl_slim_pipeline_en_3_0.html)  | 66| [ner_living_species_300_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_living_species_300_pipeline_es.html)  | 67| [ner_living_species_bert_pipeline](https://nlp.johnsnowlabs.com/2023/03/13/ner_living_species_bert_pipeline_es.html)  | 68| [ner_living_species_biobert_pipeline](https://nlp.johnsnowlabs.com/2023/03/20/ner_living_species_biobert_pipeline_en.html)  |
| 69| [ner_living_species_pipeline](https://nlp.johnsnowlabs.com/2023/03/13/ner_living_species_pipeline_en.html)  | 70| [ner_living_species_roberta_pipeline](https://nlp.johnsnowlabs.com/2023/03/13/ner_living_species_roberta_pipeline_es.html)  | 71| [ner_measurements_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_measurements_clinical_pipeline_en_3_0.html)  | 72| [ner_medication_pipeline](https://nlp.johnsnowlabs.com/2023/06/17/ner_medication_pipeline_en.html)  |
| 73| [ner_medmentions_coarse_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_medmentions_coarse_pipeline_en_3_0.html)  | 74| [ner_nature_nero_clinical_pipeline](https://nlp.johnsnowlabs.com/2023/03/14/ner_nature_nero_clinical_pipeline_en.html)  | 75| [ner_negation_uncertainty_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_negation_uncertainty_pipeline_es.html)  | 76| [ner_neoplasms_pipeline](https://nlp.johnsnowlabs.com/2023/03/15/ner_neoplasms_pipeline_es.html)  |
| 77| [ner_nihss_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_nihss_pipeline_en_3_0.html)  | 78| [ner_oncology_anatomy_general_healthcare_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_oncology_anatomy_general_healthcare_pipeline_en.html)  | 79| [ner_oncology_anatomy_general_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_oncology_anatomy_general_langtest_pipeline_en.html)  | 80| [ner_oncology_anatomy_general_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_oncology_anatomy_general_pipeline_en.html)  |
| 81| [ner_oncology_anatomy_granular_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_oncology_anatomy_granular_langtest_pipeline_en.html)  | 82| [ner_oncology_anatomy_granular_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_oncology_anatomy_granular_pipeline_en.html)  | 83| [ner_oncology_biomarker_healthcare_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_oncology_biomarker_healthcare_pipeline_en.html)  | 84| [ner_oncology_biomarker_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_oncology_biomarker_pipeline_en.html)  |
| 85| [ner_oncology_demographics_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_oncology_demographics_langtest_pipeline_en.html)  | 86| [ner_oncology_demographics_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_oncology_demographics_pipeline_en.html)  | 87| [ner_oncology_diagnosis_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_oncology_diagnosis_pipeline_en.html)  | 88| [ner_oncology_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_oncology_pipeline_en.html)  |
| 89| [ner_oncology_posology_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_oncology_posology_langtest_pipeline_en.html)  | 90| [ner_oncology_posology_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_oncology_posology_pipeline_en.html)  | 91| [ner_oncology_response_to_treatment_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_oncology_response_to_treatment_langtest_pipeline_en.html)  | 92| [ner_oncology_response_to_treatment_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_oncology_response_to_treatment_pipeline_en.html)  |
| 93| [ner_oncology_test_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_oncology_test_pipeline_en.html)  | 94| [ner_oncology_therapy_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_oncology_therapy_pipeline_en.html)  | 95| [ner_oncology_tnm_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_oncology_tnm_pipeline_en.html)  | 96| [ner_oncology_unspecific_posology_healthcare_pipeline](https://nlp.johnsnowlabs.com/2023/03/08/ner_oncology_unspecific_posology_healthcare_pipeline_en.html)  |
| 97| [ner_oncology_unspecific_posology_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_oncology_unspecific_posology_pipeline_en.html)  | 98| [ner_pathogen_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_pathogen_pipeline_en.html)  | 99| [ner_pharmacology_pipeline](https://nlp.johnsnowlabs.com/2023/03/09/ner_pharmacology_pipeline_es.html)  | 100| [ner_posology_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_posology_biobert_pipeline_en_3_0.html)  |
| 101| [ner_posology_experimental_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_posology_experimental_pipeline_en_3_0.html)  | 102| [ner_posology_greedy_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_posology_greedy_pipeline_en_3_0.html)  | 103| [ner_posology_healthcare_pipeline](https://nlp.johnsnowlabs.com/2022/03/22/ner_posology_healthcare_pipeline_en_3_0.html)  | 104| [ner_posology_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/08/ner_posology_langtest_pipeline_en.html)  |
| 105| [ner_posology_large_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_posology_large_biobert_pipeline_en_3_0.html)  | 106| [ner_posology_large_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_posology_large_pipeline_en_3_0.html)  | 107| [ner_posology_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_posology_pipeline_en_3_0.html)  | 108| [ner_posology_small_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_posology_small_pipeline_en_3_0.html)  |
| 109| [ner_radiology_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_radiology_pipeline_en_3_0.html)  | 110| [ner_radiology_wip_clinical_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_radiology_wip_clinical_pipeline_en_3_0.html)  | 111| [ner_risk_factors_biobert_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_risk_factors_biobert_pipeline_en_3_0.html)  | 112| [ner_risk_factors_pipeline](https://nlp.johnsnowlabs.com/2022/03/21/ner_risk_factors_pipeline_en_3_0.html)  |
| 113| [ner_sdoh_langtest_pipeline](https://nlp.johnsnowlabs.com/2023/09/09/ner_sdoh_langtest_pipeline_en.html)  | 114| [ner_sdoh_mentions_pipeline](https://nlp.johnsnowlabs.com/2023/06/17/ner_sdoh_mentions_pipeline_en.html)  | 115| [ner_sdoh_mentions_pipeline`](https://nlp.johnsnowlabs.com/2023/03/08/ner_sdoh_mentions_pipeline_en.html)  | 116| [ner_supplement_clinical_pipeline](https://nlp.johnsnowlabs.com/2023/03/14/ner_supplement_clinical_pipeline_en.html)  |
| 117| [ner_vop_anatomy_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_anatomy_pipeline_en.html)  | 118| [ner_vop_clinical_dept_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_clinical_dept_pipeline_en.html)  | 119| [ner_vop_demographic_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_demographic_pipeline_en.html)  | 120| [ner_vop_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_pipeline_en.html)  |
| 121| [ner_vop_problem_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_problem_pipeline_en.html)  | 122| [ner_vop_problem_reduced_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_problem_reduced_pipeline_en.html)  | 123| [ner_vop_temporal_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_temporal_pipeline_en.html)  | 124| [ner_vop_test_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_test_pipeline_en.html)  |
| 125| [ner_vop_treatment_pipeline](https://nlp.johnsnowlabs.com/2023/06/22/ner_vop_treatment_pipeline_en.html)  |

###ner_jsl_pipeline

In [ ]:
ner_pipeline = nlp.PretrainedPipeline('ner_jsl_pipeline', 'en', 'clinical/models')

ner_jsl_pipeline download started this may take some time.
Approx size to download 1.6 GB
[OK!]


In [ ]:
ner_pipeline.model.stages

[DocumentAssembler_cefadf0e0f93,
 SentenceDetectorDLModel_c83c27f46b97,
 REGEX_TOKENIZER_cef816a12166,
 WORD_EMBEDDINGS_MODEL_9004b1d00302,
 MedicalNerModel_c89cbceb1028,
 NER_CONVERTER_42a801d9e143]

In [ ]:
text = """A 28-year-old female with a history of gestational diabetes mellitus diagnosed eight years prior to presentation and subsequent type two diabetes mellitus ( T2DM ), one prior episode of HTG-induced pancreatitis three years prior to presentation , associated with an acute hepatitis , and obesity with a body mass index ( BMI ) of 33.5 kg/m2 , presented with a one-week history of polyuria , polydipsia , poor appetite , and vomiting .
Two weeks prior to presentation , she was treated with a five-day course of amoxicillin for a respiratory tract infection . She was on metformin , glipizide , and dapagliflozin for T2DM and atorvastatin and gemfibrozil for HTG . She had been on dapagliflozin for six months at the time of presentation . Physical examination on presentation was significant for dry oral mucosa ; significantly , her abdominal examination was benign with no tenderness , guarding , or rigidity .
Pertinent laboratory findings on admission were : serum glucose 111 mg/dl , bicarbonate 18 mmol/l , anion gap 20 , creatinine 0.4 mg/dL , triglycerides 508 mg/dL , total cholesterol 122 mg/dL , glycated hemoglobin ( HbA1c ) 10% , and venous pH 7.27 . Serum lipase was normal at 43 U/L . Serum acetone levels could not be assessed as blood samples kept hemolyzing due to significant lipemia .
The patient was initially admitted for starvation ketosis , as she reported poor oral intake for three days prior to admission . However , serum chemistry obtained six hours after presentation revealed her glucose was 186 mg/dL , the anion gap was still elevated at 21 , serum bicarbonate was 16 mmol/L , triglyceride level peaked at 2050 mg/dL , and lipase was 52 U/L .
The β-hydroxybutyrate level was obtained and found to be elevated at 5.29 mmol/L - the original sample was centrifuged and the chylomicron layer removed prior to analysis due to interference from turbidity caused by lipemia again . The patient was treated with an insulin drip for euDKA and HTG with a reduction in the anion gap to 13 and triglycerides to 1400 mg/dL , within 24 hours .
Her euDKA was thought to be precipitated by her respiratory tract infection in the setting of SGLT2 inhibitor use . The patient was seen by the endocrinology service and she was discharged on 40 units of insulin glargine at night , 12 units of insulin lispro with meals , and metformin 1000 mg two times a day .
It was determined that all SGLT2 inhibitors should be discontinued indefinitely . She had close follow-up with endocrinology post discharge ."""

greedy_result = ner_pipeline.fullAnnotate(text)[0]

In [ ]:
greedy_result.keys()

dict_keys(['document', 'ner_chunk', 'token', 'ner', 'embeddings', 'sentence'])

In [ ]:
import pandas as pd

chunks=[]
entities=[]
begins=[]
ends=[]

for n in greedy_result['ner_chunk']:

    chunks.append(n.result)
    begins.append(n.begin)
    ends.append(n.end)
    entities.append(n.metadata['entity'])

df = pd.DataFrame({'chunks':chunks, 'begin':begins, 'end':ends, 'entities':entities})

df

chunks  begin   end             entities
0                      28-year-old      2    12                  Age
1                           female     14    19               Gender
2    gestational diabetes mellitus     39    67             Diabetes
3                eight years prior     79    95         RelativeDate
4       type two diabetes mellitus    128   153             Diabetes
..                             ...    ...   ...                  ...
116                two times a day   2357  2371            Frequency
117               SGLT2 inhibitors   2402  2417      Drug_Ingredient
118                            She   2457  2459               Gender
119                  endocrinology   2486  2498        Clinical_Dept
120                      discharge   2505  2513  Admission_Discharge

[121 rows x 4 columns]